BaGet/Commands/RebuildSearchCommand.cs

```diff
@@ -0,0 +1,165 @@
+using BaGet.Azure.Search;
+using BaGet.Core;
+using BaGet.Protocol;
+using BaGet.Protocol.Catalog;
+using Microsoft.Azure.Search.Models;
+using Microsoft.Extensions.Logging;
+using NuGet.Packaging.Core;
+
+namespace BaGet
+{
+    public class RebuildSearchCommand : ICommand
+    {
+        private readonly NuGetClientFactory _clientFactory;
+        private readonly ICursor _cursor;
+        private readonly IPackageService _packages;
+        private readonly IndexActionBuilder _actionBuilder;
+        private readonly AzureSearchBatchIndexer _indexer;
+        private readonly ILogger<RebuildSearchCommand> _logger;
+
+        public RebuildSearchCommand(
+            NuGetClientFactory clientFactory,
+            ICursor cursor,
+            IPackageService packages,
+            IndexActionBuilder actionBuilder,
+            AzureSearchBatchIndexer indexer,
+            ILogger<RebuildSearchCommand> logger)
+        {
+            _clientFactory = clientFactory;
+            _cursor = cursor;
+            _packages = packages;
+            _actionBuilder = actionBuilder;
+            _indexer = indexer;
+            _logger = logger;
+        }
+
+        public async Task RunAsync(CancellationToken cancellationToken)
+        {
+            var minCursor = DateTimeOffset.MinValue;
+            var maxCursor = await _cursor.GetAsync(cancellationToken);
+            if (maxCursor == null)
+            {
+                maxCursor = DateTimeOffset.MinValue;
+            }
+
+            _logger.LogInformation("Finding catalog leafs committed before time {Cursor}...", maxCursor);
+
+            var catalogClient = _clientFactory.CreateCatalogClient();
+            var (catalogIndex, catalogLeafItems) = await catalogClient.LoadCatalogAsync(
+                minCursor,
+                maxCursor.Value,
+                _logger,
+                cancellationToken);
+
+            _logger.LogInformation("Removing duplicate catalog leafs...");
+
+            var packageIds = catalogLeafItems
+                .GroupBy(l => new PackageIdentity(l.PackageId, l.ParsePackageVersion()))
+                .Select(g => g.OrderByDescending(l => l.CommitTimestamp).First())
+                .Where(l => l.IsPackageDetails())
+                .Select(l => l.PackageId)
+                .Distinct(StringComparer.OrdinalIgnoreCase)
+                .ToList();
+
+            _logger.LogInformation("Processing {PackageCount} packages", packageIds.Count);
+
+            var channel = Channel.CreateBounded<IndexAction<KeyedDocument>>(new BoundedChannelOptions(5000)
+            {
+                FullMode = BoundedChannelFullMode.Wait,
+                SingleWriter = false,
+                SingleReader = false,
+            });
+
+            var produceTask = ProduceIndexActionsAsync(
+                channel.Writer,
+                new ConcurrentBag<string>(packageIds),
+                cancellationToken);
+            var consumeTask1 = ConsumeIndexActionsAsync(
```

Comment:
TODO: How many tasks does nuget.org use to push batches to Azure Search? Consider increasing concurrency

Reply:
nuget.org uses 4. Will bump to 3